### ISSUE TITLE
Error selected row not found

### ISSUE THREAD

**Issue by dpozzan** [dj-fiorex/angular2-smart-table#140]
I see this error in the console, but at the same time I don't see any limitations about the functionality of the app.

This is the complete error message: 

node_modules\pace-js\pace.min.js:6 ERROR Error: Selected row not found
    at DataSet.getSelectedRow (angular2-smart-table.mjs:863:19)
    at DataSet.select (angular2-smart-table.mjs:980:21)
    at Grid.determineRowToSelect (angular2-smart-table.mjs:1198:33)
    at Grid.processDataChange (angular2-smart-table.mjs:1172:38)
    at angular2-smart-table.mjs:1061:96
    at Object.next (Subscriber.js:110:13)
    at SafeSubscriber._next (Subscriber.js:60:26)
    at SafeSubscriber.next (Subscriber.js:31:18)
    at Subject.js:31:30
    at errorContext (errorContext.js:19:9)

-----------------------------------------------------------------------------------
**Inside class Dataset in node_modules/angular2-smart-table/lib/data-set/data-set.d.ts**

  ```
getSelectedRow() {
        if (!this.selectedRow) {
            console.warn('Selected row not found');
            throw new Error('Selected row not found');
        }
        return this.selectedRow;
    }
```

-----------------------------------------------------------------------------------
**Inside class Dataset in node_modules/angular2-smart-table/lib/data-set/data-set.d.ts**

 ```
select(selectedRowIndex) {
        if (this.getRows().length === 0) {
            throw new Error('There are no rows inside the data table');
        }
        if (this.willSelect) {
            if (this.willSelect === 'first') {
                this.selectFirstRow();
            }
            if (this.willSelect === 'last') {
                this.selectLastRow();
            }
            this.willSelect = '';
        }
        else {
            this.selectRowByIndex(selectedRowIndex);
        }
        return **this.getSelectedRow()**;
    }
```

-----------------------------------------------------------------------------------
**In class Grid in node_modules/angular2-smart-table/lib/data-source/gridt.d.ts**

 ```
/**
     * @breaking-change 1.8.0
     * Need to add `| null` in return type
     *
     * TODO: move to selectable? Separate directive
     */
    determineRowToSelect(changes) {
        if (['load', 'page', 'filter', 'sort', 'refresh'].indexOf(changes.action) !== -1) {
            return this.dataSet.**select(this.getRowIndexToSelect()**);
        }
        if (this.shouldSkipSelection()) {
            return null;
        }
        if (changes.action === 'remove') {
            if (changes.elements.length === 0) {
                // we have to store which one to select as the data will be reloaded
                this.dataSet.willSelectLastRow();
            }
            else {
                return this.dataSet.selectPreviousRow();
            }
        }
        if (changes.action === 'append') {
            // we have to store which one to select as the data will be reloaded
            this.dataSet.willSelectLastRow();
        }
        if (changes.action === 'add') {
            return this.dataSet.selectFirstRow();
        }
        if (changes.action === 'update') {
            return this.dataSet.selectFirstRow();
        }
        if (changes.action === 'prepend') {
            // we have to store which one to select as the data will be reloaded
            this.dataSet.willSelectFirstRow();
        }
        return null;
    }
```

------------------------------------------------------------------------------------------
**In class Grid in node_modules/angular2-smart-table/lib/data-source/gridt.d.ts**

 ```
processDataChange(changes) {
        if (this.shouldProcessChange(changes)) {
            this.dataSet.setData(changes.elements, this.getSelectedItems());
            if (this.getSetting('selectMode') === 'single') {
                if (this.dataSet.getRows().length > 0) {
                    const row = this.determineRowToSelect(changes);
                    this.onSelectRowSource.next(row);
                }
                else {
                    this.onSelectRowSource.next(null);
                }
            }
        }
    }
```

any suggestions ?

**Comment by uap-universe** [comment]
Can you please check which version of the smart table is in your `node_modules` folder?

The code you have posted looks a bit outdated. There was the case in a previous version where the `selectedIndex` had the value `-1` but it was not handled correctly and the code reached the `getSelectedRow()` method which threw an exception when actually no row is selected. 

But it shouldn't be the case in the version `3.0.0`.

**Comment by dpozzan** [comment]
The version of 'angular2-smart-table' in the project is '2.10.1' because we are still using Angular 15 at the moment.

I solved the problem with the error message in the console by changing the 'selectedIndex' from -1 to 0 or by deleting the setup of 'selectedIndex,' which I suppose then uses the default value of 0.

**Comment by uap-universe** [comment]
I will keep this open as long as Angular 15 has LTS and if a bug fix release for 2.10.x will be necessary, I'll pick this one up again.

I hope I will find the time to stabilize the 3.0.0 before you migrate - there are some bug tickets open I should take care of.

**Comment by LeaBroudoEcho3d** [comment]
@uap-universe My team is actually having this exact same issue—however we're on version `2.10.0`. Is there any chance you could make a bug fix for the `2.10.0` release?

In order to fix this I believe we just need to switch the ordering of the first two conditionals in `determineRowToSelect()` so that the `this.shouldSkipSelection()` check comes first. So the only change would be in `gridt.d.ts` with the following change:
```
determineRowToSelect(changes) {
       // This conditional now comes first
       if (this.shouldSkipSelection()) {
           return null;
       }

       // This conditional now comes second
       if (['load', 'page', 'filter', 'sort', 'refresh'].indexOf(changes.action) !== -1) {
           return this.dataSet.**select(this.getRowIndexToSelect()**);
       }
      ...
 }
```

**Comment by uap-universe** [comment]
@LeaBroudoEcho3d Will do probably tomorrow.
